### PR TITLE
maint(clang-tidy): Bugprone enable checks

### PR DESCRIPTION
**Pull request by Skylion007**
## Description
* Enables additional clang-tidy checks for bugprone-behavior. 

## Suggested changelog entry:

```rst
* Enable most bugprone checks in clang-tidy and fix the found potential bugs and poor coding styles.
```

### CODE REVIEW

File: .clang-tidy
Comment (Skylion007) [inline .clang-tidy:47]: We call py::print a lot in the destructors in tests which is quite noisy. Probably should nolint them all.

File: include/pybind11/numpy.h
Comment (Skylion007) [inline include/pybind11/numpy.h:1554]: @rwgk This is the prefect forwarding code that I was struggling with fixing the error for.

Comment (Skylion007) [reply include/pybind11/numpy.h]: Okay, so we do need to the copy and move constructors after all... Hmm... so we need to do some SFINAE acrobatics as described in https://moodle.ufsc.br/pluginfile.php/2377667/mod_resource/content/0/Effective_Modern_C__.pdf Item 25-26 Still struggling with it though.

**Comment by rwgk** [comment]
I just took a super quick look after seeing the emails about your last two commits. Does that mean `enable_if` isn't needed in this case, just deleting the copy and move constructors does the trick?

**Comment by Skylion007** [comment]
> 
> 
> I just took a super quick look after seeing the emails about your last two commits. Does that mean `enable_if` isn't needed in this case, just deleting the copy and move constructors does the trick?

Unfortunately, the default copy and move destructors are actually used so I do need to do the enableif trick.

File: .clang-tidy
Comment (rwgk) [reply .clang-tidy:47]: How many nolints are needed?
Are they only under tests/?

Comment (Skylion007) [reply .clang-tidy:47]: Quite a few, and yes only under tests. Under virtually ever destructor or operator that calls py::print.

Comment (rwgk) [reply .clang-tidy:47]: I'm thinking:
* We really want that check under include/
* But minimize clutter under tests/
Could we have a tests/.clang-tidy? How much trouble (copy-paste)?

Another completely different idea, possibly very naive/silly: could we make the prints noexcept?
```
void print() noexcept {
  try {
    whatever_we_have_already();
  }
  catch (...) {
    std::cout << "something useful" << std::endl;
    std::terminate();
  }
}
```

Comment (Skylion007) [reply .clang-tidy:47]: The latter could definitely work and would fundamentally solve the problem. I think the compiler's main complaint is that you may not want to call std::terminate implicitly by throwing an exception in a destructor, but calling it explicitly should be fine. 

The former doesn't because we can only specify where clang-tidy is run by build taargets (ie. cpp) and since it's a header only library we can't ask it to run on headers unless they are actually used somewhere (and therefore have their templates expanded etc.)

Comment (Skylion007) [reply .clang-tidy:47]: Alternatively, someone can just go an add all the NOLINT comments manually.

Comment (rwgk) [reply .clang-tidy:47]: > Alternatively, someone can just go an add all the NOLINT comments manually.

That sounds like a quick mostly mechanical effort, based on clang-tidy diagnostics, is that true? I'm thinking it would be best to do that straightaway in this PR, but it's fine with me to do it separately. Please let me know.

BTW: It looks like the PR description is out of date. Could you please update it to reflect the final state, if necessary?

Comment (Skylion007) [reply .clang-tidy:47]: @rwgk Let's do it separetely. Having a no-except print version with std::terminate might be a better approach longterm.

File: include/pybind11/embed.h
Comment (Skylion007) [inline include/pybind11/embed.h:192]: If we caught the exception and explicitly call std::terminate that would also work well.

File: include/pybind11/detail/common.h
Comment (Skylion007) [inline include/pybind11/detail/common.h:223]: I can revert the formatting here if necessary @rwgk . Applied by clang-format.

**Comment by Skylion007** [comment]
@henryiii Ping. Would appreciated your feedback on this PR. :)

**Comment by rwgk** [comment]
ICC flaked, I don't remember seeing that exact flake before, but it's download related, it doesn't even get to the configure stage. Maybe just try again?
(I have one suggestion though (I think). Hang on.)

File: .clang-tidy
Comment (rwgk) [reply .clang-tidy:47]: Sounds good.

File: include/pybind11/detail/common.h
Comment (rwgk) [reply include/pybind11/detail/common.h:223]: Up to you.
I often do/did, because I'm not sure I/we want to keep the current clang-format choices for preprocessor directives.

File: include/pybind11/detail/type_caster_base.h
Comment (rwgk) [inline include/pybind11/detail/type_caster_base.h:41]: This is my only strong suggestion:
IIUC we're NOT turning on bugprone-exception-escape in this PR. Is that true?
In that case it will be confusing to have the NOLINT in this PR, could you please remove them?
I realize it's a little bit of extra work adding them back in later, but IMM making this PR as straightforward/clean as possible is an important enough reason to invest those 5 minutes.

File: include/pybind11/gil.h
Comment (rwgk) [inline include/pybind11/gil.h:152]: Same here (other comment).

File: include/pybind11/numpy.h
Comment (rwgk) [inline include/pybind11/numpy.h:1554]: A 1-line hint (comment) that this is to not shadow the default copy constructor would be ideal. Maybe with a link to the clang-tidy documentation, for background.

File: tests/test_class.cpp
Comment (rwgk) [inline tests/test_class.cpp:526]: For my own benefit, merging this into the smart_holder branch, if you could minimize the whitespace changes here that would definitely be helpful. But optional.

**Review by rwgk — APPROVED** [review]

**Comment by henryiii** [comment]
Looks good, would second @rwgk's suggestions; a couple of places need comments as to why they are doing what they are doing and `bugprone-exception-escape` is probably less confusing if handled in a different PR (though personally not highly against adding a nolint for it if you know it won't pass, even if it's not enabled, but fine to wait and I do see that seeing a no lint for something that is not being linted is likely quite confusing).

**Comment by Skylion007** [comment]
@henryiii Need your approval to merge. :)

**Review by rwgk — APPROVED** [review]
Thanks Aaron!

**Comment by rwgk** [comment]
Looks like you lost a comma in numpy.h.

**Comment by rwgk** [comment]
> @henryiii Need your approval to merge. :)

We are already two maintainers.
This is pure code health.
I don't want to by-pass anyone, but to me this clearly is fine with just one approval.
Note that even in a shop like Google, where you could in theory bring down a service for billions, this is enough approvals coverage. Here we're just working on master codehealth.
Caveat: at Google we cannot submit with a broken CI. But I trust nobody would do that here.

**Pull request merged**
